File: LeetCode - Top Interview 150/103.py
# 103. Binary Tree Zigzag Level Order Traversal
# https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/


from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []

        result = []

        queue = []
        queue.append(root)

        is_even = False

        while queue:
            size = len(queue)

            current_result = [0] * size

            for i in range(size):
                node = queue.pop(0)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

                if not is_even:
                    current_result[i] = node.val
                else:
                    current_result[size - 1 - i] = node.val

            result.append(current_result)

            is_even = not is_even

        return result

File: LeetCode - Algorithm/Binary Tree/103.js
// 103. Binary Tree Zigzag Level Order Traversal
// https://leetcode.com/problems/binary-tree-zigzag-level-order-traversal/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function (root) {
  const queue = [];
  if (root) queue.push(root);

  const result = [];
  let isRight = true;

  while (queue.length > 0) {
    const currentLength = queue.length;

    const current = [];
    for (let i = 0; i < currentLength; i++) {
      const node = queue.shift();

      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);

      if (isRight) current.push(node.val);
      else current.unshift(node.val);
    }

    result.push(current);

    isRight = !isRight;
  }

  return result;
};
